test/functional/test_framework/test_framework.py

```diff
@@ -558,19 +559,21 @@ def wait_for_node_exit(self, i, timeout):
         self.nodes[i].process.wait(timeout)
 
     def connect_nodes(self, a, b):
-        from_connection = self.nodes[a]
-        to_connection = self.nodes[b]
-        ip_port = "127.0.0.1:" + str(p2p_port(b))
-        from_connection.addnode(ip_port, "onetry")
-        # poll until version handshake complete to avoid race conditions
-        # with transaction relaying
-        # See comments in net_processing:
-        # * Must have a version message before anything else
-        # * Must have a verack message before anything else
-        wait_until_helper(lambda: all(peer['version'] != 0 for peer in from_connection.getpeerinfo()))
-        wait_until_helper(lambda: all(peer['version'] != 0 for peer in to_connection.getpeerinfo()))
-        wait_until_helper(lambda: all(peer['bytesrecv_per_msg'].pop('verack', 0) == 24 for peer in from_connection.getpeerinfo()))
-        wait_until_helper(lambda: all(peer['bytesrecv_per_msg'].pop('verack', 0) == 24 for peer in to_connection.getpeerinfo()))
+        id_a = self.nodes[a].p2p_get_peer_id()
+        id_b = self.nodes[b].p2p_get_peer_id()
+
+        addr_a = p2p_url(a) + "/p2p/" + id_a
+        ret = self.nodes[b].p2p_connect(addr_a)
```

Comment:
The connect call is synchronous, right? So it'll be definitely connected when the function returns. Or is it async such that it will attempt to connect?

I can imagine it's synchonous now. But in the future, connecting to peers should be done asynchronously, where requesting to connect basically means "we add the node to the list of nodes we want to connect to". However, this will require having different peer types, which I think isn't implemented right now.

Reply:
Btw, once things are async, we have to use the `wait_until_helper` from above.

Reply:
Yeah its synchronous. Either an error is thrown or the communication succeeds.